src/main/scala/scala/collection/next/NextIterableOnceOpsExtensions.scala

```diff
@@ -0,0 +1,101 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.collection
+package next
+
+import scala.language.implicitConversions
+
+private[next] final class NextIterableOnceOpsExtensions[A, CC[_], C](
+  private val col: IterableOnceOps[A, CC, C]
+) extends AnyVal {
+  import NextIterableOnceOpsExtensions.GroupMap
+
+  /**
+   * Partitions this IterableOnce into a map according to a discriminator function `key`. All the values that
+   * have the same discriminator are then transformed by the `value` function and then reduced into a
+   * single value with the `reduce` function.
+   *
+   * {{{
+   *   def occurrences[A](as: IterableOnce[A]): Map[A, Int] =
+   *     as.iterator.groupMapReduce(identity)(_ => 1)(_ + _)
+   * }}}
+   *
+   * @note This will force the evaluation of the Iterator.
+   */
+  def groupMapReduce[K, B](key: A => K)(f: A => B)(reduce: (B, B) => B): immutable.Map[K, B] =
+    groupMapTo(key)(f).reduce(reduce)
+
+  def groupByTo[K](key: A => K): GroupMap[A, K, A, immutable.Iterable, immutable.Map] =
+    groupMapTo(key)(identity)
+
+  def groupMapTo[K, V](key: A => K)(f: A => V): GroupMap[A, K, V, immutable.Iterable, immutable.Map] =
+    new GroupMap(col, key, f, immutable.Iterable, immutable.Map)
+}
+
+object NextIterableOnceOpsExtensions {
+  final case class GroupMap[A, K, V, CC[_], MC[_, _]](
```

Comment:
I'm not super thrilled with how many copies you end up needing to do if you want to change everything. I kind of think it would be better if it was a mutable type instead where the original type is no longer useable after calling a method on it (similar to how calling a method on `Iterator` makes the previous `Iterator` unusable)

Reply:
Yeah, I actually do not want it to be a **case class** because the `copy` would be public and it could change things which shouldn't be changed; but it was just for a quick prototype.

Now, I had the same doubt. For example what if someone one calls `result` and then `reduce`? If the original collection was strict like **List** it will work; but if not, like **Iterator**, it will fail. So, I thought that maybe the best is to say that such usage is discouraged... However, keeping it immutable could open the door to it be reusable a shared across functions; but but but, I am not sure if I like or dislike that.

What do you think?

Reply:
I personally think that an immutable builder is what I would prefer dealing with (ie, I don’t want to have to think about whether it is safe to alias it or not). I don’t think the cost of copies is significant compared to the final group by operation.
But I think we should not use a case class, just in case we want to add other fields in the future, which would be impossible to achieve with case classes without breaking binary compatibility.